### PR TITLE
CSHARP-409 - Mapper: Support timestamp at CqlQueryOptions level (Inserts only for now)

### PR DESCRIPTION
**Pull request by collinsauve**
From https://datastax-oss.atlassian.net/browse/CSHARP-409

### CODE REVIEW

File: src/Cassandra/Mapping/CqlBatch.cs [hunk 1]
```diff
@@ -77,7 +77,7 @@ private void Insert<T>(bool ifNotExists, bool insertNulls, T poco, CqlQueryOptio
             var values = getBindValues(poco);
             //generate INSERT query based on null values (if insertNulls set)
             object[] queryParameters;
-            var cql = _cqlGenerator.GenerateInsert<T>(insertNulls, values, out queryParameters, ifNotExists, ttl);
+            var cql = _cqlGenerator.GenerateInsert<T>(insertNulls, values, out queryParameters, ifNotExists, ttl, queryOptions?.Timestamp);
```
Comment (jorgebay) [inline src/Cassandra/Mapping/CqlBatch.cs]: Null-conditional operator is a C# 6 language feature. Can you replace it with conditional operator?

We are still using C# 4.0, [will start using C# 5 soon](https://datastax-oss.atlassian.net/browse/CSHARP-426) but using C# 6.0 means requiring newer compiler versions (dropping msbuild 12).

**Comment by collinsauve** [comment]
@jorgebay fixed.  Thanks.

File: src/Cassandra/Mapping/Mapper.cs [hunk 2]
```diff
@@ -211,7 +211,7 @@ public Task InsertAsync<T>(T poco, bool insertNulls, int? ttl, CqlQueryOptions q
             var values = getBindValues(poco);
             object[] queryParameters;
             //generate INSERT query based on null values (if insertNulls set)
-            var cql = _cqlGenerator.GenerateInsert<T>(insertNulls, values, out queryParameters, ttl: ttl);
+            var cql = _cqlGenerator.GenerateInsert<T>(insertNulls, values, out queryParameters, ttl: ttl, timestamp: queryOptions?.Timestamp);
```
Comment (jorgebay) [inline src/Cassandra/Mapping/Mapper.cs]: Another one :)

**Comment by collinsauve** [comment]
Oops I missed that one.  Fixed.

Also see PR #208: Using lang features > C# 4.0 should be a compile error.

**Comment by jorgebay** [comment]
Looks good, I will try to add a unit test for it and merge it.

**Comment by jorgebay** [comment]
After a closer look, I'm not going to merge it as it should be implemented using protocol level timestamp: you can specify it at `Statement` level without affecting the query.

You can implementing by adding it to the following method:
https://github.com/datastax/csharp-driver/blob/a4afd10b15d33bde2606d657f0f59f59fb539389/src/Cassandra/Mapping/CqlQueryOptions.cs#L131

Something like:

```csharp
if (Timestamp != null)
{
    statement.SetTimestamp(Timestamp);
}
```

That later gets applied by the statement factory here: https://github.com/datastax/csharp-driver/blob/a4afd10b15d33bde2606d657f0f59f59fb539389/src/Cassandra/Mapping/Statements/StatementFactory.cs#L44

We appreciate your continuing effort to contribute and improve the driver, keep it coming!

Sorry for the misunderstanding

**Comment by collinsauve** [comment]
Implemented according to your suggestions at PR #209

**Pull request closed without merging**